---
title: "Marker_specific_plots"
author: "Sajal Sthapit"
date: '2022-07-07'
output: html_document
---
Draw marker specific plots
```{r Define functions}
rm(list = ls())
library(tidyverse)
library(cowplot)
# FUNCTION DEFINITIONS ----

custom_box_plot <- function(inp, sub.text = ""){
  # inp is a tibble that includes a column for y-axis variable (Year), x-axis variable (selected marker), and category variable columns for faceting
  # example inp: inp <- select(samples, marker$SNPid[[1]], Year, Habit, Region, MC)
  x.var <- names(inp[1])
  y.var <- names(inp[2])
  facet.var <- names(inp[3])
  p <- ggplot(inp, aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ inp[[facet.var]]) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, subtitle = sub.text ) 
  return(p)
}

custom_bar_plot <- function(inp, sub.text = ""){
  # inp is a tibble with 3 columns: first column is the category of variety, second column is the allelic state for the marker (column name), third column is the count n: e.g., Habit | IWB653314 | n. This table is generated by count(input, Habit, IWB57868)
  # Extract category and marker names
  category <- names(inp[1])
  alleles <- names(inp[2])
  count <- names(inp[3])
  
  # Plots
  out <-  ggplot(inp, aes_string(x = alleles, y = count)) +  
    geom_bar(stat = "identity") +
    facet_grid(reformulate(category)) + # use reformulate to pass the string in category as ~ variable
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, y = "Count", subtitle = sub.text)
return(out)

}

draw_combined_custom_plots <- function(inp, marker.snpid, marker.details = marker.snpid){
  # master function that draws all plots, but the results can be cluttered.
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  inp2 <- mutate(inp, Region = "All regions") # do this internally so that only one dataframe is given to the function.
  inp3 <- mutate(inp, Habit = "both")
  
  p.box <- list(habit = rbind(inp, inp3) %>% select(., marker.snpid, Year, Habit) %>% custom_box_plot(sub.text = "Habit"),
                # region-level plots
                both.region = filter(inp, Region %in% c("Eastern", "Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  select(., marker.snpid,  Year, Region) %>% custom_box_plot(sub.text = "All varieties by region"),
                spr.region = filter(inp, Habit == "spring", Region %in% c("Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  select(., marker.snpid, Year, Region) %>% custom_box_plot(sub.text = "Spring varieties by region"),
                win.region = filter(inp, Habit == "winter", Region %in% c("Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  select(., marker.snpid, Year, Region) %>% custom_box_plot(sub.text = "Winter varieties by region"),
                # state-level plots
                both.state = filter(inp, state.code %in% 
                # included states with a mix of both habits, with one habit at least 10% of total
                                      c("CO", "ID", "MN", "MT", "ND",
                                        "NE", "OR", "SD", "UT", "WA", 
                                        "WI")) %>% 
                  select(., marker.snpid, Year, Region_State) %>% custom_box_plot(sub.text = "All varieties by state"),
                spr.state = filter(inp, Habit == "spring", state.code %in%
                # included states with 6 or more spring varieties
                                      c("AZ", "CA", "CO", "ID", "MN",
                                       "MT", "ND", "NE", "OR", "SD", "WA")) %>% 
                  select(., marker.snpid, Year, Region_State) %>% custom_box_plot(sub.text = "Spring varieties by state"),
                win.state = filter(inp, Habit == "winter", state.code %in%
                                     c("CO", "ID", "IL", "IN", "KS", 
                                       "MT", "NE", "NY", "OH", "OK", 
                                       "OR", "SC", "TX", "VA", "WA")) %>%
                  select(., marker.snpid, Year, Region_State) %>% 
                  custom_box_plot(sub.text = "Winter varieties by state")
                  )
  
  p.bar <- list(habit = rbind(inp, inp3) %>% count(., Habit, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Habit"),
                both.region = filter(inp, Region %in% c("Eastern", "Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  count(., Region, .[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by region"),
                spr.region = filter(inp, Habit == "spring", Region %in% c("Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  count(., Region, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Spring varieties by region"),
                win.region = filter(inp, Habit == "winter", Region %in% c("Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  count(., Region, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Winter varieties by region"),
                # state-level plots
                both.state = filter(inp, state.code %in% 
                # included states with a mix of both habits, with one habit at least 10% of total
                                      c("CO", "ID", "MN", "MT", "ND",
                                        "NE", "OR", "SD", "UT", "WA", 
                                        "WI")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by state"),
                spr.state = filter(inp, Habit == "spring", state.code %in%
                # included states with 6 or more spring varieties
                                      c("AZ", "CA", "CO", "ID", "MN",
                                       "MT", "ND", "NE", "OR", "SD", "WA")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Spring varieties by state"),
                win.state = filter(inp, Habit == "winter", state.code %in%
                                     c("CO", "ID", "IL", "IN", "KS", 
                                       "MT", "NE", "NY", "OH", "OK", 
                                       "OR", "SC", "TX", "VA", "WA")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Winter varieties by state"),
                #BP and decades
                both.bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by BP"),
                spr.bp = filter(inp, Habit == "spring") %>% count(., BP, .[marker.snpid]) %>% 
                  custom_bar_plot(sub.text = "Spring varieties by BP"),
                win.bp = filter(inp, Habit == "winter") %>% count(., BP, .[marker.snpid]) %>% 
                  custom_bar_plot(sub.text = "Winter varieties by BP"),
                both.decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by decade"),
                spr.decade = filter(inp, Habit == "spring") %>% count(., Decade, .[marker.snpid]) %>% 
                  custom_bar_plot(sub.text = "Spring varieties by decade"),
                win.decade = filter(inp, Habit == "winter") %>% count(., Decade, .[marker.snpid]) %>% 
                  custom_bar_plot("Winter varieties by decade")
                )
  out <- list(
    #both
    row01 = plot_grid(p.box$habit, p.box$both.region, nrow = 1, rel_widths = c(.6, 1)),
    row02 = plot_grid(p.bar$habit, p.bar$both.region, nrow = 1, rel_widths = c(.6, 1)),
    row03 = p.box$both.state,
    row04 = p.bar$both.state,
    row05 = plot_grid(p.bar$both.bp, p.bar$both.decade, nrow = 1, rel_widths = c(.3, 1)),
    #spring
    row06 = plot_grid(p.box$spr.region, p.bar$spr.region, nrow = 1, rel_widths = c(1, 1)),
    row07 = p.box$spr.state,
    row08 = p.bar$spr.state,
    row09 = plot_grid(p.bar$spr.bp, p.bar$spr.decade, nrow = 1, rel_widths = c(.3, 1)),
    #winter
    row10 = plot_grid(p.box$win.region, p.bar$win.region, nrow = 1, rel_widths = c(1, 1)),
    row11 = p.box$win.state,
    row12 = p.bar$win.state,
    row13 = plot_grid(p.bar$win.bp, p.bar$win.decade, nrow = 1, rel_widths = c(.3, 1))
    )

  return(plot_grid(out$row01, out$row02, out$row03, out$row04, out$row05,
                   out$row06, out$row07, out$row08, out$row09,
                   out$row10, out$row11, out$row12, out$row13,
                   ncol = 1, nrow = 13))
}

draw_combined_custom_plots2 <- function(inp, marker.snpid, marker.details = marker.snpid){
  # same as the master plot above but excludes BP and Decade to reduce clutter.
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  inp2 <- mutate(inp, Region = "All regions") # do this internally so that only one dataframe is given to the function.
  inp3 <- mutate(inp, Habit = "both")
  
  p.box <- list(habit = rbind(inp, inp3) %>% select(., marker.snpid, Year, Habit) %>% custom_box_plot(sub.text = "Habit"),
                # region-level plots
                both.region = filter(inp, Region %in% c("Eastern", "Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  select(., marker.snpid,  Year, Region) %>% custom_box_plot(sub.text = "All varieties by region"),
                spr.region = filter(inp, Habit == "spring", Region %in% c("Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  select(., marker.snpid, Year, Region) %>% custom_box_plot(sub.text = "Spring varieties by region"),
                win.region = filter(inp, Habit == "winter", Region %in% c("Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  select(., marker.snpid, Year, Region) %>% custom_box_plot(sub.text = "Winter varieties by region"),
                # state-level plots
                both.state = filter(inp, state.code %in% 
                # included states with a mix of both habits, with one habit at least 10% of total
                                      c("CO", "ID", "MN", "MT", "ND",
                                        "NE", "OR", "SD", "UT", "WA", 
                                        "WI")) %>% 
                  select(., marker.snpid, Year, Region_State) %>% custom_box_plot(sub.text = "All varieties by state"),
                spr.state = filter(inp, Habit == "spring", state.code %in%
                # included states with 6 or more spring varieties
                                      c("AZ", "CA", "CO", "ID", "MN",
                                       "MT", "ND", "NE", "OR", "SD", "WA")) %>% 
                  select(., marker.snpid, Year, Region_State) %>% custom_box_plot(sub.text = "Spring varieties by state"),
                win.state = filter(inp, Habit == "winter", state.code %in%
                                     c("CO", "ID", "IL", "IN", "KS", 
                                       "MT", "NE", "NY", "OH", "OK", 
                                       "OR", "SC", "TX", "VA", "WA")) %>%
                  select(., marker.snpid, Year, Region_State) %>% 
                  custom_box_plot(sub.text = "Winter varieties by state")
                  )
  
  p.bar <- list(habit = rbind(inp, inp3) %>% count(., Habit, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Habit"),
                both.region = filter(inp, Region %in% c("Eastern", "Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  count(., Region, .[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by region"),
                spr.region = filter(inp, Habit == "spring", Region %in% c("Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  count(., Region, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Spring varieties by region"),
                win.region = filter(inp, Habit == "winter", Region %in% c("Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  count(., Region, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Winter varieties by region"),
                # state-level plots
                both.state = filter(inp, state.code %in% 
                # included states with a mix of both habits, with one habit at least 10% of total
                                      c("CO", "ID", "MN", "MT", "ND",
                                        "NE", "OR", "SD", "UT", "WA", 
                                        "WI")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by state"),
                spr.state = filter(inp, Habit == "spring", state.code %in%
                # included states with 6 or more spring varieties
                                      c("AZ", "CA", "CO", "ID", "MN",
                                       "MT", "ND", "NE", "OR", "SD", "WA")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Spring varieties by state"),
                win.state = filter(inp, Habit == "winter", state.code %in%
                                     c("CO", "ID", "IL", "IN", "KS", 
                                       "MT", "NE", "NY", "OH", "OK", 
                                       "OR", "SC", "TX", "VA", "WA")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% custom_bar_plot(sub.text = "Winter varieties by state")
                #BP and decades
                #both.bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by BP"),
                #both.decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by decade")
                )
  out <- list(
    #both
    row01 = plot_grid(p.box$habit, p.box$both.region, nrow = 1, rel_widths = c(.6, 1)),
    row02 = plot_grid(p.bar$habit, p.bar$both.region, nrow = 1, rel_widths = c(.6, 1)),
    row03 = p.box$both.state,
    row04 = p.bar$both.state,
    #spring
    row05 = plot_grid(p.box$spr.region, p.bar$spr.region, nrow = 1, rel_widths = c(1, 1)),
    row06 = p.box$spr.state,
    row07 = p.bar$spr.state,
    #winter
    row08 = plot_grid(p.box$win.region, p.bar$win.region, nrow = 1, rel_widths = c(1, 1)),
    row09 = p.box$win.state,
    row10 = p.bar$win.state
    )

  return(plot_grid(out$row01, out$row02, out$row03, out$row04, 
                   out$row05, out$row06, out$row07, 
                   out$row08, out$row09, out$row10,
                   ncol = 1, nrow = 10))
}

draw_combined_custom_plots_bp <- function(inp, marker.snpid, marker.details = marker.snpid){
  # function only draws breeding period and decade plots. drawing these separately to reduce clutter on the main plots
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  inp2 <- mutate(inp, Region = "All regions") # do this internally so that only one dataframe is given to the function.
  inp3 <- mutate(inp, Habit = "both")

  p.bar <- list(#BP and decades
                both.bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by BP"),
                spr.bp = filter(inp, Habit == "spring") %>% count(., BP, .[marker.snpid]) %>% 
                  custom_bar_plot(sub.text = "Spring varieties by BP"),
                win.bp = filter(inp, Habit == "winter") %>% count(., BP, .[marker.snpid]) %>% 
                  custom_bar_plot(sub.text = "Winter varieties by BP"),
                both.decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot(sub.text = "All varieties by decade"),
                spr.decade = filter(inp, Habit == "spring") %>% count(., Decade, .[marker.snpid]) %>% 
                  custom_bar_plot(sub.text = "Spring varieties by decade"),
                win.decade = filter(inp, Habit == "winter") %>% count(., Decade, .[marker.snpid]) %>% 
                  custom_bar_plot("Winter varieties by decade")
                )
  out <- list(
    #both
    both = plot_grid(p.bar$both.bp, p.bar$both.decade, nrow = 1, rel_widths = c(.3, 1)),
    #spring
    spr = plot_grid(p.bar$spr.bp, p.bar$spr.decade, nrow = 1, rel_widths = c(.3, 1)),
    #winter
    win = plot_grid(p.bar$win.bp, p.bar$win.decade, nrow = 1, rel_widths = c(.3, 1))
    )

  return(plot_grid(out$both, out$spr, out$win, ncol = 1, nrow = 3))
}
```

```{r Draw plots, message=FALSE, warning=FALSE}
g.nuc <- read_tsv("data/genotype_nucleotide_format_10391_loci_imputed.txt")
trans.g <- g.nuc[7:ncol(g.nuc)] %>% t %>% as_tibble()
names(trans.g) <- g.nuc$SNPid
samples <- read_tsv("data/variety_details.txt")
sum(samples$GS.Sample.ID == names(g.nuc[7:ncol(g.nuc)])) # sample name orders match
input <- cbind(select(samples, Corrected.Sample.ID, ACNO, Habit, Region, State, Year, MC, Decade, BP), trans.g)
state.code <- tibble(State = unique(input$State) %>% sort)
region.code <- tibble(Region    = c("Eastern", "Great Plains", "Northern", "Pacific", "PNW", "AK"),
                      region.code  = c("EAS", "GPL", "NOR", "PAC", "PNW", "AK"))
state.code <- left_join(state.code, distinct(samples[c("State", "Region")]), by = "State")
state.code <- left_join(state.code, region.code, by = "Region")
state.code$State
state.code <- mutate(state.code, state.code = c("AK", "AZ", "AR", "CA", "CO", "DE", "FL",
                                          "GA", "ID", "IL", "IN", "IO", "KS", "KY",
                                          "LA", "MD", "MI", "MN", "MO", "MT", "NE",
                                          "NM", "NY", "NC", "ND", "OH", "OK", "OR",
                                          "PA", "SC", "SD", "TN", "TX", "UT", "VT",
                                          "VA", "WA", "WV", "WI"))
state.code <- mutate(state.code, Region_State = str_c(region.code, ":", state.code))
input <- left_join(input, select(state.code, -Region), by = "State") %>% 
  select(., region.code, state.code, Region_State, everything())
table(input$Habit, input$state.code)
# states with more than 20 varieties: CA, CO, ID, IN KS, MN, MT, NE, NY, ND, OR, SC, TX, WA
# states with more than 9 varieties: AZ, CA, CO, ID, IL, IN, KS, MN, MT, NE, NY, HD, OH, OK, OR, SC, SD, TX, UT, VA, WA, WI

start <- Sys.time()
map(.x = g.nuc$SNPid,
    .f = ~ select(input, .x, Year, Habit, Region, State, BP, Decade, state.code, Region_State) %>% 
      draw_combined_custom_plots2(., marker.snpid = .x) %>% 
      ggsave(filename = str_c(.x, ".png"), device = "png", units = "px", width = 2000, height = 3800, dpi = 200))

map(.x = g.nuc$SNPid[7952:10391],
    .f = ~ select(input, .x, Year, Habit, Region, State, BP, Decade, state.code, Region_State) %>% 
      draw_combined_custom_plots_bp(., marker.snpid = .x) %>% 
      ggsave(filename = str_c(.x, "_bp.png"), device = "png", units = "px", width = 2000, height = 1333, dpi = 200))
end <- Sys.time()
end - start # run time for drawing plots
```
