---
title: "Calculations of Rsb and xpEHH"
author: "Sajal Sthapit"
date: "6/21/2022"
output: html_document
---
# RSB and XP-EHH calculations
06/03/2021
Rsb is the natural log of the ratio of integrated EHHS of pop1 over pop2. If there is selection in pop1 then inES(pop1)>inES(pop2) and therefore Rsb value is extreme +ve. Selection in pop2 will be extreme -ve value. Same applies for xpEHH (even stated in Sabeti 2007). Therefore, it is important to note what were assigned as population 1 and 2.
Ignore sign in calc_candidate_regions allows you to select only +ve values, hence my target population should be set as pop1 not pop2 as I had done previously.

```{r message=FALSE, warning=FALSE}
rm(list = ls())
source("functions/functions_for_selection_sweep_analysis.R") # load custom functions for the analysis
library(tidyverse)
library(rehh)
library(pegas) # for Fst calculation
geno <- list(nuc = read_tsv("data/genotype_nucleotide_format_10391_loci_imputed.txt"),
             ab  = read_tsv("data/genotype_AB_format_10391_loci_imputed.txt"))
res <- readRDS("output/scan_hh_ihs_results_polFALSE_sgap2.5MB_mgapNAMB_discardBorderTRUE.rds")
```
# Allele frequencies
Allele frequencies within all populations are calculated and saved as a data frame within population name.

```{r calculate allele frequencies}
allele.freq <- map(.x = names(res$pop), .f = ~ calc_allele_freq(pop.name = .x))
names(allele.freq) <- names(res$pop)
```

# Define population pairs and calculate Rsb, xpEHH, and Fst
```{r message=FALSE, warning=FALSE}
# define pop.pairs in script. Target population for selection is in population 1.
pop.pairs <- tibble(pop1 = c("both.all2", "both.CA", "both.CO", "both.eas", "both.eas2", 
                             "both.gpl", "both.gpl2", "both.ID", "both.IN", "both.KS",
                             "both.MN", "both.MT", "both.ND", "both.NE", "both.nor", 
                             "both.nor2", "both.pac", "both.pac2", "both.pnw",
                             "both.pnw2", "both.SC", "both.TX", "both.WA",
                             #spring
                             "spr.all", "spr.CA", "spr.gpl", "spr.ID", "spr.MN", 
                             "spr.ND", "spr.nor", "spr.nor2", "spr.pac",
                             "spr.pac2", "spr.pnw", "spr.pnw2", "spr.spr2", "spr.WA",
                             #winter
                             "win.all", "win.CO", "win.eas", "win.eas2", "win.gpl", 
                             "win.gpl2", "win.ID", "win.IN", "win.KS", "win.NE",
                             "win.nor", "win.NY", "win.OR", "win.pnw", "win.pnw2", 
                             "win.SC", "win.WA", "win.win2"),
                    pop2 = c("both.all1", "both.CA_Other", "both.CO_Other", "both.eas_Other", "both.eas1",
                             "both.gpl_Other", "both.gpl1", "both.ID_Other", "both.IN_Other", "both.KS_Other",
                             "both.MN_Other", "both.MT_Other", "both.ND_Other", "both.NE_Other", "both.nor_Other",
                             "both.nor1", "both.pac_Other", "both.pac1", "both.pnw_Other",
                             "both.pnw1", "both.SC_Other", "both.TX_Other", "both.WA_Other",
                             #spring
                             "win.all", "spr.CA_Other", "spr.gpl_Other", "spr.ID_Other", "spr.MN_Other",
                             "spr.ND_Other", "spr.nor_Other", "spr.nor1", "spr.pac_Other",
                             "spr.pac1", "spr.pnw_Other", "spr.pnw1", "spr.spr1", "spr.WA_Other",
                             #winter
                             "spr.all", "win.CO_Other", "win.eas_Other", "win.eas1", "win.gpl_Other", 
                             "win.gpl1", "win.ID_Other", "win.IN_Other", "win.KS_Other", "win.NE_Other",
                             "win.nor_Other", "win.NY_Other", "win.OR_Other", "win.pnw_Other", "win.pnw1", 
                             "win.SC_Other", "win.WA_Other", "win.win1"))

# Calculate RSB, XP-EHH, and Fst ----
rsb <- map2(.x = pop.pairs$pop1, .y = pop.pairs$pop2,
            .f = ~ ines2rsb(scan_pop1 = res[[.x]][["scan"]],
                           scan_pop2 = res[[.y]][["scan"]], include_freq = TRUE) )
xpe <- map2(.x = pop.pairs$pop1, .y = pop.pairs$pop2,
            .f = ~ ies2xpehh(scan_pop1 = res[[.x]][["scan"]],
                           scan_pop2 = res[[.y]][["scan"]], include_freq = TRUE) )

names(rsb) <- pop.pairs$pop1
names(xpe) <- pop.pairs$pop1

#Fst calculations
geno$diploid <- geno$nuc[7:ncol(geno$nuc)]
geno$diploid[geno$diploid == "A"] <- "A/A"
geno$diploid[geno$diploid == "C"] <- "C/C"
geno$diploid[geno$diploid == "G"] <- "G/G"
geno$diploid[geno$diploid == "T"] <- "T/T"
colnames(geno$diploid) <- colnames(geno$nuc[7:ncol(geno$nuc)]) # restore column names
fst <- map2(.x = pop.pairs$pop1, .y = pop.pairs$pop2,
            .f = ~ calc_fst(pop1 = .x, pop2 = .y))
names(fst) <- pop.pairs$pop1

# Save results
results.to.save <- list(pop = res$pop, chr.pos = res$chr.pos, pop.pairs = pop.pairs,
                        snp.gap.percentile = res$snp.gap.percentile, snp.gap.quantile = res$snp.gap.quantile, 
                        scan_hh.parameters = res$scan_hh.parameters,
                        fst = fst, rsb = rsb, xpe = xpe, allele.freq = allele.freq)
saveRDS(object = results.to.save, file = "output/allele_freq_Fst_Rsb_xpEHH.Rds")
```
