---
title: "Marker_specific_plots"
author: "Sajal Sthapit"
date: '2022-07-07'
output: html_document
---
Draw marker specific plots for spring varieties only
```{r}
rm(list = ls())
library(tidyverse)
library(cowplot)
g.nuc <- read_tsv("data/genotype_nucleotide_format_10391_loci_imputed.txt")
trans.g <- g.nuc[7:ncol(g.nuc)] %>% t %>% as_tibble()
names(trans.g) <- g.nuc$SNPid
samples <- read_tsv("data/variety_details.txt")
sum(samples$GS.Sample.ID == names(g.nuc[7:ncol(g.nuc)])) # sample name orders match
input <- cbind(select(samples, Corrected.Sample.ID, ACNO, Habit, Region, State, Year, MC, Decade, BP), trans.g) %>% 
  filter(Habit == "spring") # only subset spring varieties
table(input$State)
# states with 20 or more varieties: California, Idaho, Minnesota, North Dakota, Washington
# states with 9 or more varieties: c("Arizona", "California")
start <- Sys.time()
map(.x = g.nuc$SNPid,
    .f = ~ select(input, .x, Year, Habit, Region, State, BP, Decade) %>% 
      draw_combined_custom_plots(., marker.snpid = .x) %>% 
      ggsave(filename = str_c(.x, ".png"), device = "png", units = "px", width = 3000, height = 2000))
end <- Sys.time()
end - start # run time

# FUNCTION DEFINITIONS ----

custom_box_plot <- function(inp){
  # inp is a tibble that includes a column for y-axis variable (Year), x-axis variable (selected marker), and category variable columns for faceting
  # example inp: inp <- select(samples, marker$SNPid[[1]], Year, Habit, Region, MC)
  x.var <- names(inp[1])
  y.var <- names(inp[2])
  facet.var <- names(inp[3])
  p <- ggplot(inp, aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ inp[[facet.var]]) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, subtitle = facet.var)
  return(p)
}

custom_bar_plot <- function(inp){
  # inp is a tibble with 3 columns: first column is the category of variety, second column is the allelic state for the marker (column name), third column is the count n: e.g., Habit | IWB653314 | n. This table is generated by count(input, Habit, IWB57868)
  # Extract category and marker names
  category <- names(inp[1])
  alleles <- names(inp[2])
  count <- names(inp[3])
  
  # Plots
  out <-  ggplot(inp, aes_string(x = alleles, y = count)) +  
    geom_bar(stat = "identity") +
    facet_grid(reformulate(category)) + # use reformulate to pass the string in category as ~ variable
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, y = "Count", subtitle = category)
return(out)

}

draw_combined_custom_plots <- function(inp, marker.snpid, marker.details = marker.snpid){
  # function modified to use with spring population
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  inp2 <- mutate(inp, Region = "All regions") # do this internally so that only one dataframe is given to the function.
  
  p.box <- list(region = rbind(inp, inp2) %>% # combine the two inputs
                  select(., marker.snpid,  Year, Region) %>% 
                  filter(Region %in% c("All regions", "Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  custom_box_plot()
                  )
  p.bar <- list(region = rbind(inp, inp2) %>% 
                  count(., Region, .[marker.snpid]) %>% 
                  filter(Region %in% c("All regions", "Great Plains", "Northern", "Pacific", "PNW")) %>% 
                  custom_bar_plot(),
                state = filter(inp, State %in% c("Arizona", "California", "Colorado", "Idaho", "Minnesota", 
                                                 "Montana", "North Dakota", "Oregon", "South Dakota", "Washington")) %>% 
                  count(., State, .[marker.snpid]) %>% 
                  custom_bar_plot(),
                bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(),
                decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot()
                )
  out <- list(
    row1 = plot_grid(p.box$region, p.bar$region, nrow = 1, rel_widths = c(1, 1)),
    row2 = plot_grid(p.bar$bp, p.bar$decade, nrow = 1, rel_widths = c(.3, 1))
    )

  return(plot_grid(out$row1, p.bar$state, out$row2, ncol = 1, nrow = 3))
}

draw_combined_custom_plots(inp = input, marker.snpid = "IWB57868")

```

Old function definitions
```{r}
# FUNCTION DEFINITIONS ----
# This function extracts the marker with the most extreme iHS, Rsb, xpEHH, or Fst value from each candidate region. The list output generated can then be used as input to generate the custom bar plots.
extract_best_markers <- function(cr.df, scan.res, stat.label){ 
  # cr.df = data.frame of candidate regions with CHR, START, END, and Category columns.
  # scan.res = results of iHS, Rsb, xpEHH, or Fst calculation respectively.
  # stat.label = name of the statistics column in the scan.res (IHS, RSB, or XPEHH). Note that the Fst values for fst.scan.res is also stored in the column names RSB to be compatible for the rehh::calc_candidate_regions function.
  pops.with.res <- intersect(names(scan.res), cr.df$Category) # populations that returned candidate regions
  inp <- filter(cr.df, Category %in% pops.with.res) # why are some candidate regions not in the ihs object? The should match.
  marker <- list()
  marker$SNPid <- map(.x = 1:nrow(inp), 
      .f = ~ scan.res[inp$Category][[1]] %>% # isolates the right data.frame with the population-specific scan results
        filter(CHR == inp$CHR[[.x]], POSITION >= inp$START[[.x]], POSITION <= inp$END[[.x]]) %>% 
        arrange(desc(abs(.data[[stat.label]]))) %>% 
        .[["SNPid"]] %>% .[[1]]    ) # works
  marker$details <- map(.x = 1:nrow(inp), 
      .f = ~ scan.res[inp$Category][[1]] %>% # isolates the right data.frame with the population-specific scan results
        filter(CHR == inp$CHR[[.x]], POSITION >= inp$START[[.x]], POSITION <= inp$END[[.x]]) %>% 
        arrange(desc(abs(.data[[stat.label]]))) %>% 
        slice_head(n = 1) %>% 
        mutate(details = str_c(SNPid, " at ", CHR, ":", POSITION/1e6, " MBp")) %>% 
          .[["details"]]    ) # works
 return(marker) 
}

new_custom_box_plot <- function(x, y, category){
  # inp is a tibble that includes a column for y-axis variable (Year), x-axis variable (selected marker), and category variable columns for faceting
  # example inp: inp <- select(samples, marker$SNPid[[1]], Year, Habit, Region, MC)
  x.var <- names(inp[1])
  y.var <- names(inp[2])
  
  p.habit <- ggplot(inp, aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ Habit) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() +
    labs(x = NULL, subtitle = "Habit")
  p.region <- filter(inp, Region != "AK") %>% # Remove Alaska as it only has 4 varieties
    ggplot(., aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ Region) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() +
    labs(x = NULL, subtitle = "Region")
  p.mc <- filter(inp, !MC %in% c("HWS", "HWW", "xxW")) %>% # remove market classes with few varieties
    ggplot(., aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ MC) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() +
    labs(x = NULL, subtitle = "Class")
  
  p.plots <- plot_grid(p.habit, p.region, p.mc, nrow = 1, rel_widths = c(.5, 1, 1))
  p.label <- ggdraw() + draw_label(str_c("Alleles for ", x.axis.label))
  return(plot_grid(p.label, p.plots, ncol = 1, rel_heights = c(0.1, 1)))
}

custom_box_plot <- function(inp, x.axis.label = NULL){
  # inp is a tibble that includes a column for y-axis variable (Year), x-axis variable (selected marker), and category variable columns for faceting
  # example inp: inp <- select(samples, marker$SNPid[[1]], Year, Habit, Region, MC)
  x.var <- names(inp[1])
  y.var <- names(inp[2])
  
  p.habit <- ggplot(inp, aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ Habit) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() +
    labs(x = NULL, subtitle = "Habit")
  p.region <- filter(inp, Region != "AK") %>% # Remove Alaska as it only has 4 varieties
    ggplot(., aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ Region) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() +
    labs(x = NULL, subtitle = "Region")
  p.mc <- filter(inp, !MC %in% c("HWS", "HWW", "xxW")) %>% # remove market classes with few varieties
    ggplot(., aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ MC) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() +
    labs(x = NULL, subtitle = "Class")
  
  p.plots <- plot_grid(p.habit, p.region, p.mc, nrow = 1, rel_widths = c(.5, 1, 1))
  p.label <- ggdraw() + draw_label(str_c("Alleles for ", x.axis.label))
  return(plot_grid(p.label, p.plots, ncol = 1, rel_heights = c(0.1, 1)))
}

custom_bar_plot <- function(inp, x.axis.label = NULL){
  # inp is a tibble with 3 columns: first column is the category of variety, second column is the allelic state for the marker (column name), third column is the count n: e.g., Habit | IWB653314 | n. This table is generated by count(input, Habit, IWB57868)
  # Extract category and marker names
  category <- names(inp[1])
  alleles <- names(inp[2])
  count <- names(inp[3])
  
  # Plots
  out <-  ggplot(inp, aes_string(x = alleles, y = count)) +  
    geom_bar(stat = "identity") +
    facet_grid(reformulate(category)) + # use reformulate to pass the string in category as ~ variable
    theme_bw() +
    labs(x = x.axis.label, y = "Count", subtitle = category)
return(out)

}

draw_combined_custom_plots <- function(inp, marker.snpid, marker.details = marker.snpid){
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  out.box <- select(inp, marker.snpid, Year, Habit, Region, MC) %>% custom_box_plot(., marker.details)
  # combine bar plots
  p.bar <- list(habit = count(inp, Habit, inp[marker.snpid]) %>% custom_bar_plot(),
                region = count(inp, Region, inp[marker.snpid]) %>% custom_bar_plot(),
                mc = count(inp, MC, inp[marker.snpid]) %>% custom_bar_plot(),
                bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(),
                decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot(),
                row1 = plot_grid(p.bar$habit, p.bar$region, p.bar$mc, nrow = 1, rel_widths = c(.4, 1, 1)),
                row2 = plot_grid(p.bar$bp, p.bar$decade, nrow = 1, rel_widths = c(.3, 1))
                )
  p.bar$habit <- count(inp, Habit, inp[marker.snpid]) %>% custom_bar_plot()
  p.bar$region <- table("Region" = inp$Region, "alleles" = inp[[marker.snpid]]) %>% as_tibble %>% 
    filter(Region != "AK") %>%  custom_bar_plot()
  p.bar$mc <- table("Class" = inp$MC, "alleles" = inp[[marker.snpid]]) %>% as_tibble %>% 
    filter(!Class %in% c("HWS", "HWW", "xxW")) %>%  custom_bar_plot()
  p.bar$bp <- table("Period" = inp$BP, "alleles" = inp[[marker.snpid]]) %>% as_tibble %>% custom_bar_plot()
  p.bar$decade <- table("Decade" = inp$Decade, "alleles" = inp[[marker.snpid]]) %>% as_tibble %>% custom_bar_plot()
  p.bar$row1 <- plot_grid(p.bar$habit, p.bar$region, p.bar$mc, nrow = 1, rel_widths = c(.4, 1, 1))
  p.bar$row2 <- plot_grid(p.bar$bp, p.bar$decade, nrow = 1, rel_widths = c(.3, 1))
  out.bar <- plot_grid(p.bar$row1, p.bar$row2)
  return(plot_grid(out.box, out.bar, ncol = 1))
}

p.bar

draw_combined_custom_plots(inp = input, marker.snpid = "IWB50666")

# APPLY functions ----
marker <- extract_best_markers(cr.df = cr.rsb.df, scan.res = rsb, stat.label = "RSB")

draw_combined_custom_plots(samples, marker$SNPid[[1]], marker$details[[1]])
map(.x = 1:3,
    .f = ~ draw_combined_custom_plots(inp = samples, marker$SNPid[[.x]], marker$details[[.x]]))
```