---
title: "Marker_specific_plots"
author: "Sajal Sthapit"
date: '2022-07-07'
output: html_document
---
Draw marker specific plots
```{r Define functions}
rm(list = ls())
library(tidyverse)
library(cowplot)
# FUNCTION DEFINITIONS ----

custom_box_plot <- function(inp){
  # inp is a tibble that includes a column for y-axis variable (Year), x-axis variable (selected marker), and category variable columns for faceting
  # example inp: inp <- select(samples, marker$SNPid[[1]], Year, Habit, Region, MC)
  x.var <- names(inp[1])
  y.var <- names(inp[2])
  facet.var <- names(inp[3])
  p <- ggplot(inp, aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ inp[[facet.var]]) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, subtitle = facet.var)
  return(p)
}

custom_bar_plot <- function(inp){
  # inp is a tibble with 3 columns: first column is the category of variety, second column is the allelic state for the marker (column name), third column is the count n: e.g., Habit | IWB653314 | n. This table is generated by count(input, Habit, IWB57868)
  # Extract category and marker names
  category <- names(inp[1])
  alleles <- names(inp[2])
  count <- names(inp[3])
  
  # Plots
  out <-  ggplot(inp, aes_string(x = alleles, y = count)) +  
    geom_bar(stat = "identity") +
    facet_grid(reformulate(category)) + # use reformulate to pass the string in category as ~ variable
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, y = "Count", subtitle = category)
return(out)

}

draw_combined_custom_plots <- function(inp, marker.snpid, marker.details = marker.snpid){
  # function modified to use with spring population
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  inp2 <- mutate(inp, Region = "All regions") # do this internally so that only one dataframe is given to the function.
  
  p.box <- list(region = rbind(inp, inp2) %>% # combine the two inputs
                  select(., marker.snpid,  Year, Region) %>% 
                  filter(Region %in% c("All regions", "Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  custom_box_plot()
                  )
  p.bar <- list(region = rbind(inp, inp2) %>% 
                  count(., Region, .[marker.snpid]) %>% 
                  filter(Region %in% c("All regions", "Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  custom_bar_plot(),
                state = filter(inp, State %in% c("Colorado", "Idaho", "Illinois", "Indiana", "Kansas", 
                                                 "Montana", "Nebraska", "New York", "Ohio", "Oklahoma", 
                                                 "Oregon", "South Carolina", "Texas", "Virginia", "Washington")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% 
                  custom_bar_plot(),
                bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(),
                decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot()
                )
  out <- list(
    row1 = plot_grid(p.box$region, p.bar$region, nrow = 1, rel_widths = c(1, 1)),
    row2 = plot_grid(p.bar$bp, p.bar$decade, nrow = 1, rel_widths = c(.3, 1))
    )

  return(plot_grid(out$row1, p.bar$state, out$row2, ncol = 1, nrow = 3))
}

draw_combined_custom_plots2 <- function(inp, marker.snpid, marker.details = marker.snpid){
  # function modified to use with spring population
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  inp2 <- mutate(inp, Region = "All regions") # do this internally so that only one dataframe is given to the function.
  
  p.box <- list(region = rbind(inp, inp2) %>% # combine the two inputs
                  select(., marker.snpid,  Year, Region) %>% 
                  filter(Region %in% c("All regions", "Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  custom_box_plot(),
                state = filter(inp, State %in% c("Colorado", "Idaho", "Illinois", "Indiana", "Kansas", 
                                                 "Montana", "Nebraska", "New York", "Ohio", "Oklahoma", 
                                                 "Oregon", "South Carolina", "Texas", "Virginia", "Washington")) %>% 
                        select(., marker.snpid, Year, Region_State) %>% 
                        custom_box_plot()
                  )
  p.bar <- list(region = rbind(inp, inp2) %>% 
                  count(., Region, .[marker.snpid]) %>% 
                  filter(Region %in% c("All regions", "Eastern", "Great Plains", "Northern", "PNW")) %>% 
                  custom_bar_plot(),
                state = filter(inp, State %in% c("Colorado", "Idaho", "Illinois", "Indiana", "Kansas", 
                                                 "Montana", "Nebraska", "New York", "Ohio", "Oklahoma", 
                                                 "Oregon", "South Carolina", "Texas", "Virginia", "Washington")) %>% 
                  count(., Region_State, .[marker.snpid]) %>% 
                  custom_bar_plot(),
                bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(),
                decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot()
                )
  out <- list(
    row1 = plot_grid(p.box$region, p.bar$region, nrow = 1, rel_widths = c(1, 1)),
    row2 = plot_grid(p.bar$bp, p.bar$decade, nrow = 1, rel_widths = c(.3, 1))
    )

  return(plot_grid(out$row1, p.box$state, p.bar$state, out$row2, ncol = 1, nrow = 4))
}
```

```{r Draw plots}
g.nuc <- read_tsv("data/genotype_nucleotide_format_10391_loci_imputed.txt")
trans.g <- g.nuc[7:ncol(g.nuc)] %>% t %>% as_tibble()
names(trans.g) <- g.nuc$SNPid
samples <- read_tsv("data/variety_details.txt")
sum(samples$GS.Sample.ID == names(g.nuc[7:ncol(g.nuc)])) # sample name orders match
input <- cbind(select(samples, Corrected.Sample.ID, ACNO, Habit, Region, State, Year, MC, Decade, BP), trans.g) %>% 
  filter(Habit == "winter") # only subset winter varieties
table(input$State)
# states with 20 or more varieties: CO, ID, IN, KS, NE, NY, OR, SC, WA
# states with 9 or more varieties: c("Colorado", "Idaho", "Illinois", "Indiana", "Kansas", "Montana", "Nebraska", "New York", "Ohio", "Oklahoma", "Oregon", "South Carolina", "Texas", "Virginia", "Washington")
state.code <- tibble(State = unique(input$State) %>% sort)
state.code <- mutate(state.code, Code = c("AR", "CA", "CO", "DE", "FL", "GA", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "MD", "MI", "MN",
                                          "MO", "MT", "NE", "NM", "NY", "NC", "ND", "OH", "OK", "OR", "PA", "SC", "SD", "TN", "TX", "UT",
                                          "VT", "VA", "WA", "WV", "WI"))
state.code <- mutate(state.code, Region_State = c("EAS:AR", "PAC:CA", "GPL:CO", "EAS:DE", "EAS:FL", "EAS:GA", "PNW:ID", "EAS:IL",
                                                  "EAS:IN", "EAS:IA", "GPL:KS", "EAS:KY", "EAS:LA", "EAS:MD", "EAS:MI", "NOR:MN",
                                                  "EAS:MO", "NOR:MT", "GPL:NE", "GPL:NM", "EAS:NY", "EAS:NC", "NOR:ND", "EAS:OH", 
                                                  "GPL:OK", "PNW:OR", "EAS:PA", "NOR:SC", "NOR:SD", "EAS:TN", "GPL:TX", "PNW:UT",
                                                  "EAS:VT", "EAS:VA", "PNW:WA", "EAS:WV", "EAS:WI"))
input <- left_join(input, state.code, by = "State") %>% 
  select(., Code, Region_State, everything())
start <- Sys.time()
map(.x = g.nuc$SNPid[1],
    .f = ~ select(input, .x, Year, Habit, Region, State, BP, Decade, Code, Region_State) %>% 
      draw_combined_custom_plots2(., marker.snpid = .x) %>% 
      ggsave(filename = str_c(.x, ".png"), device = "png", units = "px", width = 3200, height = 3200))
end <- Sys.time()
end - start # run time for drawing plots
```
