---
title: "Marker_specific_plots"
author: "Sajal Sthapit"
date: '2022-07-07'
output: html_document
---
Draw marker specific plots
```{r}
rm(list = ls())
library(tidyverse)
library(cowplot)

# FUNCTION DEFINITIONS ----
# This function extracts the marker with the most extreme iHS, Rsb, xpEHH, or Fst value from each candidate region. The list output generated can then be used as input to generate the custom bar plots.
custom_box_plot <- function(inp){
  # inp is a tibble that includes a column for y-axis variable (Year), x-axis variable (selected marker), and category variable columns for faceting
  # example inp: inp <- select(samples, marker$SNPid[[1]], Year, Habit, Region, MC)
  x.var <- names(inp[1])
  y.var <- names(inp[2])
  facet.var <- names(inp[3])
  p <- ggplot(inp, aes_string(x = x.var, y = y.var)) +
    geom_violin() + geom_boxplot(alpha = 0.5) +
    facet_grid(. ~ inp[[facet.var]]) +
    stat_summary(fun = mean, geom = "point", shape = 15, size = 3, color = "black") +
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, subtitle = facet.var)
  return(p)
}

custom_bar_plot <- function(inp){
  # inp is a tibble with 3 columns: first column is the category of variety, second column is the allelic state for the marker (column name), third column is the count n: e.g., Habit | IWB653314 | n. This table is generated by count(input, Habit, IWB57868)
  # Extract category and marker names
  category <- names(inp[1])
  alleles <- names(inp[2])
  count <- names(inp[3])
  
  # Plots
  out <-  ggplot(inp, aes_string(x = alleles, y = count)) +  
    geom_bar(stat = "identity") +
    facet_grid(reformulate(category)) + # use reformulate to pass the string in category as ~ variable
    theme_bw() + theme(panel.grid = element_blank()) +
    labs(x = NULL, y = "Count", subtitle = category)
return(out)

}

draw_combined_custom_plots <- function(inp, marker.snpid, marker.details = marker.snpid){
  # inp is a genotype data frame with category columns (Habit, Region, etc.) and all the markers as columns
  # marker.snpid is the SNPid to be plotted
  # marker.details is the descriptive label to use for the plot (e.g., "IWB61100 at 1A:482 MBp")
  
  p.box <- list(habit = select(inp, marker.snpid, Year, Habit) %>% custom_box_plot(),
                  region = select(inp, marker.snpid,  Year, Region) %>% 
                  filter(Region != "AK") %>% custom_box_plot()
                  )
  out.box <- plot_grid(p.box$habit, p.box$region, nrow = 1, rel_widths = c(.5, 1))
  # combine bar plots
  p.bar <- list(habit = count(inp, Habit, inp[marker.snpid]) %>% custom_bar_plot(),
                region = count(inp, Region, inp[marker.snpid]) %>% 
                  filter(Region != "AK") %>% custom_bar_plot(),
                bp = count(inp, BP, inp[marker.snpid]) %>% custom_bar_plot(),
                decade = count(inp, Decade, inp[marker.snpid]) %>% custom_bar_plot()
                )
  out.bar <- list(
    row1 = plot_grid(p.bar$habit, p.bar$region, nrow = 1, rel_widths = c(.5, 1)),
    row2 = plot_grid(p.bar$bp, p.bar$decade, nrow = 1, rel_widths = c(.3, 1))
    )

  return(plot_grid(out.box, out.bar$row1, out.bar$row2, ncol = 1, nrow = 3))
}
```

```{r draw plots}
g.nuc <- read_tsv("data/genotype_nucleotide_format_10391_loci_imputed.txt")
trans.g <- g.nuc[7:ncol(g.nuc)] %>% t %>% as_tibble()
names(trans.g) <- g.nuc$SNPid
samples <- read_tsv("data/variety_details.txt")
sum(samples$GS.Sample.ID == names(g.nuc[7:ncol(g.nuc)])) # sample name orders match
input <- cbind(select(samples, Corrected.Sample.ID, ACNO, Habit, Region, State, Year, MC, Decade, BP), trans.g)
table(input$State)
# states with more than 20 varieties: CA, CO, ID, IN KS, MN, MT, NE, NY, ND, OR, SC, TX, WA
# states with more than 9 varieties: AZ, CA, CO, ID, IL, IN, KS, MN, MT, NE, NY, HD, OH, OK, OR, SC, SD, TX, UT, VA, WA, WI
state.code <- tibble(State = unique(input$State) %>% sort)
region.code <- tibble(Region    = c("Eastern", "Great Plains", "Northern", "Pacific", "PNW", "AK"),
                      reg.code  = c("EAS", "GPL", "NOR", "PAC", "PNW", "AK"))
state.code <- left_join(state.code, distinct(samples[c("State", "Region")]), by = "State")
state.code <- left_join(state.code, region.code, by = "Region")
state.code$State
state.code <- mutate(state.code, Code = c("AK", "AZ", "AR", "CA", "CO", "DE", "FL",
                                          "GA", "ID", "IL", "IN", "IO", "KS", "KY",
                                          "LA", "MD", "MI", "MN", "MO", "MT", "NE",
                                          "NM", "NY", "NC", "ND", "OH", "OK", "OR",
                                          "PA", "SC", "SD", "TN", "TX", "UT", "VT",
                                          "VA", "WA", "WV", "WI"))
state.code <- mutate(state.code, Region_State = str_c(reg.code, ":", Code))

# Global analysis
start <- Sys.time()
map(.x = g.nuc$SNPid[1:2],
    .f = ~ select(input, .x, Year, Habit, Region, BP, Decade) %>% 
      draw_combined_custom_plots(., marker.snpid = .x) %>% 
      ggsave(filename = str_c(.x, ".png"), device = "png", units = "px", width = 3200, height = 2000))
end <- Sys.time()
end - start # run time

```